### ISSUE TITLE
op scaled_dot_product_attention case different results

### ISSUE THREAD

**Issue by Dengshunge** [pytorch/pytorch#112015]
### 🐛 Describe the bug

when I set the `dropout_p=0.0`, the result is different.  But `dropout_p=-1`, the result is same. Maybe the op scaled_dot_product_attention has some bug. Please fix it, thank you.
```
import os
import torch
import numpy as np
import cv2
import random
import torch.nn.functional as F

seed = 0
random.seed(seed)
np.random.seed(seed)
torch.manual_seed(seed)
torch.cuda.manual_seed(seed)
torch.cuda.manual_seed_all(seed)
os.environ['PYTHONHASHSEED'] = str(seed)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

query = torch.rand((2, 8, 4096, 40)).float().cuda()
key = torch.rand((2, 8, 4096, 40)).float().cuda()
value = torch.rand((2, 8, 4096, 40)).float().cuda()

hidden_states1 = F.scaled_dot_product_attention(
            query, key, value, attn_mask=None, dropout_p=-1, is_causal=False
        ) 
hidden_states2 = F.scaled_dot_product_attention(
            query, key, value, attn_mask=None, dropout_p=-1, is_causal=False
        ) 
print((hidden_states1-hidden_states2).sum())

### Versions

PyTorch version: 2.0.1+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.27.6
Libc version: glibc-2.27

Python version: 3.10.13 (main, Sep 11 2023, 13:44:35) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.0-144-generic-x86_64-with-glibc2.27
Is CUDA available: True
CUDA runtime version: 11.5.119
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060
Nvidia driver version: 495.29.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
架构：           x86_64
CPU 运行模式：   32-bit, 64-bit
字节序：         Little Endian
CPU:             12
在线 CPU 列表：  0-11
每个核的线程数： 2
每个座的核数：   6
座：             1
NUMA 节点：      1
厂商 ID：        GenuineIntel
CPU 系列：       6
型号：           165
型号名称：       Intel(R) Core(TM) i5-10400F CPU @ 2.90GHz
步进：           3
CPU MHz：        3069.540
CPU 最大 MHz：   4300.0000
CPU 最小 MHz：   800.0000
BogoMIPS：       5799.77
虚拟化：         VT-x
L1d 缓存：       32K
L1i 缓存：       32K
L2 缓存：        256K
L3 缓存：        12288K
NUMA 节点0 CPU： 0-11
标记：           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp md_clear flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.26.0
[pip3] torch==2.0.1
[pip3] torchaudio==2.0.2
[pip3] torchvision==0.15.2
[pip3] triton==2.0.0
[conda] numpy                     1.26.0                   pypi_0    pypi
[conda] torch                     2.0.1                    pypi_0    pypi
[conda] torchaudio                2.0.2                    pypi_0    pypi
[conda] torchvision               0.15.2                   pypi_0    pypi
[conda] triton                    2.0.0                    pypi_0    pypi

**Comment by justinchuby** [comment]
When dropout is 0.5 the output is not deterministic. Was that not what you expected?

**Comment by Dengshunge** [comment]
In theory, when dropout great than 0, the output should not be deterministic. However, in this code, when dropout equal 0, the output also is not deterministic. 
@justinchuby

**Comment by syusukexrj** [comment]
same problem. centos + torch 2.0.1 + cu117

**Comment by Dinghow** [comment]
sama problem

**Comment by bobqianic** [comment]
same problem
